### PR TITLE
Add a Replace proposal

### LINKED ISSUE #170: Add a Replace proposal
### Description of feature:

The proposal here is to add a Replace proposal of the following form:

```
struct {
    uint32 leaf_index;
    LeafNode leaf_node;
} Replace;
```

The semantic of this proposal would be the same as an Update from the member at `leaf_index` with the indicated `leaf_node`.  In particular, the proposal validation rules and identity continuity rules would apply in the same way, mutatis mutandis.  But of course, it could be sent by someone other than the updating member.

@mulmarta notes that this mechanism could be used to "roll back" a member to an earlier (possibly compromised) state.  We could mitigate this risk by adding a `leaf_node_version` extension carrying a `uint32` counter and requiring that Updates/Replaces increase this counter.

### Use case:

MLS proposals have to be committed in the epoch in which they are sent, but the asynchronous nature of MLS operations means that the Commit advancing from epoch N to epoch N+1 might not include some proposals sent in epoch N.  For Add and Remove proposals, the committer can "re-originate" the proposals as inline proposals with the commit.  For Update, however, this is not possible, since the location of the replaced leaf is determined by the sender of the proposal.  This slows down PCS, since the sender might have to re-send an Update, potentially multiple times.

### Implementation discussion (Optional)

The implementation of this feature should parallel the implementation of the Update proposal.  I have begun hacking a little bit, and aside from understanding the whole flow of proposal generation, validation, and application, the main challenging thing I have seen so far is managing the caching of the secrets corresponding to a LeafNode.  Right now, since we know an Update will never be applied after the Commit for its epoch is received, we can erase the cached secrets when advancing from the epoch.  Here, the lifetime of the secrets can span epochs, so we would need some story for how they are deleted.

### PR DESCRIPTION
**Pull request by bifurcation**
### Issues:

Resolves #170 

### Description of changes:

This PR implements a Replace proposal long the lines discussed in #170 and [draft-barnes-mls-replace](https://datatracker.ietf.org/doc/draft-barnes-mls-replace/).  Changes are gated behind a feature flag `replace_proposal`, as discussed in #171.

### Call-outs:

There are a few questions around error codes highlighted with `XXX(RLB)` in the code.

The current code does not reflect support for the Replace proposal in Capabilities when this feature is enabled, and probably should to be specification-compliant.  Somewhat surprisingly, this does not cause tests to fail.  I suspect there's a general bug to fix in that there doesn't seem to be any check that a proposal is supported by the group before it is applied.

### Testing:

In general, the testing strategy is broadly parallel to testing of Update, since this new proposal type is closely related:

* Tests that the commit builder properly includes Replace proposals in `src/group/commit.rs`
* Tests of overall functionality in `src/group/mod.rs`
* Tests that proposal list rules are applied in `src/group/proposal_cache.rs`

As a small drive-by, I also added tests that duplicate Updates are correctly filtered / detected, and the corresponding code to implement this policy.

I think these tests are complete and in the right places, but my familiarity with this code base is pretty rudimentary.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT license.

### CODE REVIEW

File: mls-rs/src/group/mod.rs [hunk 1]
```diff
@@ -745,34 +755,57 @@ where
...
+
+            let mut updated_leaves = updated_leaves;
+            let self_update = updated_leaves.next();
+
+            if updated_leaves.count() > 0 {
+                // XXX(RLB) This error code is wrong; might need a new value?  Or just not bother
+                // checking, assuming checks have been done upstream?  But then we might also need a
+                // new error code there.  In any case, we need to enforce that there are not multiple
+                // Update **or Replace** proposals for the same leaf.
+                return Err(MlsError::InvalidCommitSelfUpdate);
```
Comment (mulmarta) [inline mls-rs/src/group/mod.rs:797]: Didn't we already made sure of that when validating the proposal list? So is this check redundant?

File: mls-rs/src/group/mod.rs [hunk 2]
```diff
@@ -908,32 +941,84 @@ where
...
+            leaf_node,
+        }))
+    }
+
+    /// Abandon any cached state corresponding to a leaf node
+    #[cfg(feature = "replace_proposal")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn abandon_leaf_node(&mut self, leaf_node: &LeafNode) {
+        self.pending_updates.remove(&leaf_node.public_key);
+    }
```
Comment (mulmarta) [inline mls-rs/src/group/mod.rs:982]: Isn't it annoying for the application to track  `LeafNode`'s?

Can we make it automatic so that leaves with epoch <= current leaf epoch are deleted automatically since they can no longer be applied? Storing other leaves can't hurt security (since they haven't been applied) so this function isn't so crucial anymore.

Also, how about a function that deletes all cached leaves, or all leaves with `not_before < given time T`, or all leaves from epoch `< given epoch E`?

**Review by mulmarta — COMMENTED** [review]
Thanks @bifurcation !

I'm still reviewing, but I added some initial comments I had when looking.

File: mls-rs-crypto-cryptokit/src/lib.rs [hunk 3]
```diff
@@ -21,6 +21,19 @@ use mls_rs_core::{
...
+pub enum UniFfiCompatibleError {
+    #[error("oh no")]
+    OhNo,
+}
+
+#[uniffi::export(with_foreign)]
+pub trait BoxedConstant: Send + Sync {
+    fn add(&self, b: u32) -> Result<u32, UniFfiCompatibleError>;
+}
+
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs-crypto-cryptokit/src/lib.rs:36]: Shouldn't be.  Accidentally committed.  (Investigating using UniFFI to call Swift crypto functions.)

File: mls-rs/src/group/proposal_cache.rs [hunk 4]
```diff
@@ -2593,6 +2646,128 @@ mod tests {
...
+                .cache(update2_ref.clone(), update2.clone(), bob)
+                .send()
+                .await
+                .unwrap();
+
+        assert_eq!(processed_proposals.0.len(), 1);
+        assert_eq!(processed_proposals.1.unused_proposals.len(), 1);
+
+        // Proposals are processed in an unpredictable order, so we can't test that a specific
+        // proposal was selected.  We just check that one was selected and the other was rejected.
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs/src/group/proposal_cache.rs:2700]: Yeah, I didn't bother to figure out why it was unpredictable, but removing that variability would make this test nicer.

[the remark this answers is not in the document]
Comment (mulmarta) [reply mls-rs/src/group/proposal_cache.rs:2700]: It's because proposals are cached in a HashMap which uses a randomized hash

File: mls-rs/src/group/snapshot.rs [hunk 5]
```diff
@@ -37,8 +40,8 @@ pub(crate) struct Snapshot {
     private_tree: TreeKemPrivate,
     epoch_secrets: EpochSecrets,
     key_schedule: KeySchedule,
-    #[cfg(feature = "by_ref_proposal")]
-    pending_updates: SmallMap<HpkePublicKey, (HpkeSecretKey, Option<SignatureSecretKey>)>,
+    #[cfg(any(feature = "by_ref_proposal", feature = "replace_proposal"))]
+    pending_updates: SmallMap<HpkePublicKey, PendingUpdate>,
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs/src/group/snapshot.rs:44]: It depends on how tightly you think Replace and Update are.  If you think that the LeafNode in a Replace could come from anywhere as long as it has the right form, then they're more independent.  If you think that it ought to come only from an Update, then they're more tightly coupled.  I could probably be convinced of the latter.

File: mls-rs/src/group/proposal_filter/filtering.rs [hunk 6]
```diff
@@ -449,6 +614,17 @@ pub(crate) fn proposer_can_propose(
...
+                | ProposalType::UPDATE
+                | ProposalType::REPLACE
+                | ProposalType::REMOVE
+                | ProposalType::PSK
+                | ProposalType::RE_INIT
+                | ProposalType::GROUP_CONTEXT_EXTENSIONS
+        ),
         #[cfg(feature = "by_ref_proposal")]
         (Sender::External(_), false) => false,
         #[cfg(feature = "by_ref_proposal")]
```
[the remark this answers is not in the document]
Comment (mulmarta) [reply mls-rs/src/group/proposal_filter/filtering.rs:630]: Easy to mix up :)

File: mls-rs/src/group/snapshot.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mulmarta) [reply mls-rs/src/group/snapshot.rs:44]: AFAIR, a leaf in Replace must have `LeafNodeSource = Update` so where else can it come from? Maybe it's OK to be strict?

File: mls-rs/src/group/mod.rs [hunk 7]
```diff
@@ -1801,8 +1886,8 @@ where
         #[cfg(feature = "by_ref_proposal")]
         self.state.proposals.clear();
 
-        // Clear the pending updates list
-        #[cfg(feature = "by_ref_proposal")]
+        // Clear the pending updates list (but only if they can't show up in a Replace)
+        #[cfg(all(feature = "by_ref_proposal", not(feature = "replace_proposal")))]
         {
             self.pending_updates = Default::default();
         }
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs/src/group/mod.rs:1884]: Actually:
1. LeafNodes in Commits never end up in this cache
2. We should delete the state entry for a LeafNode that is committed via Update or Replace is deleted when that proposal is applied.  I had thought this PR added that, but it doesn't seem to.  I'll make the change to do this.

With (2), you actually get the same GC effect as here as long as everything gets committed.

[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs/src/group/mod.rs:1884]: On trying to implement, it seems that deleting on epoch change is the right answer.  I have updated that code so that if Replace is allowed, we delete:

1. Any state corresponding to the current leaf node (which has already been used)
2. Any state that was added in an epoch earlier than the current leaf node (which will be stale)

File: mls-rs/src/group/proposal_filter/filtering.rs [hunk 8]
```diff
@@ -220,6 +232,72 @@ where
             proposals.update_senders.remove(i);
         });
 
+        // Check that Replaces are valid
+        #[cfg(feature = "replace_proposal")]
+        {
+            let bad_indices: Vec<_> = wrap_iter(proposals.replace_proposals())
```
[the remark this answers is not in the document]
Comment (bifurcation) [reply mls-rs/src/group/proposal_filter/filtering.rs:241]: What would this look like?

File: mls-rs/src/client_builder.rs [hunk 9]
```diff
@@ -879,8 +882,14 @@ pub(crate) struct Settings {
 
 impl Default for Settings {
     fn default() -> Self {
+        #[cfg(not(feature = "replace_proposal"))]
+        let extension_types = Default::default();
+
+        #[cfg(feature = "replace_proposal")]
+        let extension_types = vec![ExtensionType::LEAF_NODE_EPOCH];
```
Comment (tomleavy) [inline mls-rs/src/client_builder.rs:889]: I'm curious if this should be default or not just because the feature is on? I think I view the feature more as "you can possibly do this" vs "you definitely want to do this"

File: mls-rs/src/group/commit.rs [hunk 10]
```diff
@@ -197,6 +197,19 @@ where
...
     }
 
+    /// Insert a [`ReplaceProposal`](crate::group::proposal::ReplaceProposal) into
+    /// the current commit that is being built.
+    #[cfg(feature = "replace_proposal")]
+    pub fn replace_member(
+        mut self,
+        to_replace: u32,
+        leaf_node: LeafNode,
+    ) -> Result<Self, MlsError> {
```
[the remark this answers is not in the document]
Comment (tomleavy) [reply mls-rs/src/group/commit.rs:207]: Agreed, to date we have made LeafNode private, maybe it should stay private?

File: mls-rs/src/group/mod.rs [hunk 11]
```diff
@@ -265,8 +280,7 @@ where
     private_tree: TreeKemPrivate,
     key_schedule: KeySchedule,
     #[cfg(feature = "by_ref_proposal")]
-    pending_updates:
-        crate::map::SmallMap<HpkePublicKey, (HpkeSecretKey, Option<SignatureSecretKey>)>, // Hash of leaf node hpke public key to secret key
+    pending_updates: SmallMap<HpkePublicKey, PendingUpdate>, // Hash of leaf node hpke public key to secret key
```
Comment (tomleavy) [inline mls-rs/src/group/mod.rs:283]: @mulmarta I _think_ that we persist this data to disk as part of the group state itself? If so making this change would break loading group states from before this version. We can probably just try to deserialize the new format followed by the old format somehow.
